Don't autosave unchanged has_one through records

alan opened the pull request:
The bug is with the autosave logic. It was calling save on all has_one through records in the callbacks so that meant that the after_commit callback gets called on the has_one through records even though they don't participate in the transaction!

chanks committed:
Roll back open transactions when the running thread is killed.

lsylvester committed:
add implementation of respond_to? for ActiveSupport::Duration

agis committed:
Time#change can now change nanoseconds (:nsec)

Closes #16392.

alan commented:
@Lolecule want to double check? Opened a PR to my own repo so that it could reviewed before sending it to Rails.

cc/ @greysteil @jamienglish

SteveParrington commented:
Yeah it looks good to me, hopefully they'll get it merged soon!

english commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -1116,6 +1119,22 @@ def test_should_not_load_the_associated_model
+    MemberDetail.create(organization: organization, member: @member)
+  end
+
+  def test_should_not_save_organization
```
The other test cases don't seem to mention the specifics of the instances, but rather the general purpose of the test, so maybe something along the lines of `test_should_not_save_has_one_through_model` would fit better?

alan replied:
Good catch!

aantix committed:
Changed the DangerousAttributeError exception message to include messaging that the conflict could be because of a conflicting attribute.

Commit:
Changes puts to logger.info

senny committed:
reuse view test-cases for pg materialized view tests.

senny committed:
pg, add test cases for updateable views.

rafaelfranca committed:
Merge pull request #16874 from Bartuz/patch-1

[docs] Changes puts to logger.info [ci skip]

seuros committed:
Bump globalid version

tanmay3011 committed:
[ci skip] #find_in_batches doesn't support :include option

matthewd committed:
Merge pull request #16893 from tanmay3011/docs

[ci skip] #find_in_batches doesn't support :include option

seuros committed:
lock suckerpunch version to 1.1

jeremy committed:
Merge pull request #16895 from seuros/suck1-1

lock suckerpunch version to 1.1

seuros committed:
Revert "lock suckerpunch version to 1.1"

kostia committed:
MessageVerifier raises an appropriate exception if the secret is nil


Otherwise this will lead to another error later on 
when generating a signature:
TypeError (no implicit conversion of nil into String).

brandonhilkert committed:
Remove version. No longer using Celluloid 0.16.

spastorino committed:
Merge pull request #16898 from brandonhilkert/master

Remove version. No longer using Celluloid 0.16.

spastorino committed:
Merge pull request #16896 from seuros/revert-16895-suck1-1

Revert "lock suckerpunch version to 1.1"

rafaelfranca committed:
Merge pull request #16890 from seuros/globalid

Bump globalid version

rafaelfranca committed:
Use latest globalid

rafaelfranca committed:
Globalid was released

kostia committed:
Changes "if secret.nil?" to unless secret in MessageVerfier

rafaelfranca committed:
Merge pull request #16897 from kostia/message-varifier-raises-exception-on-nil-secret

MessageVerifier.new raises an appropriate exception if the secret is nil

rafaelfranca committed:
Change gid calls to to_gid

rafaelfranca committed:
Se the test order of activejob tests

rafaelfranca committed:
Prepare maintenance policy for 4.2 release [ci skip]

rafaelfranca committed:
Merge pull request #16669 from aantix/dangerous_attribute_error_better_message

Improved the error messaging for the DangerousAttributeError exception

y-yagi committed:
Change ActiveJob test directory to "test/jobs"

seuros committed:
Merge pull request #16908 from y-yagi/change_activejob_test_dir

Change ActiveJob test directory to "test/jobs" [ci skip]

kuldeepaggarwal committed:
use `allow_blank` option instead

matthewd committed:
Revert "Merge pull request #15312 from JuanitoFatas/action_view/asset_path"

This reverts commit 21ec7fefea5a1b823cb88733b019b3d52240801d.

Per #16911, there's actually a good reason for a "blank" value to have a
useful `to_s` here. So let's also add some tests to prove that.

matthewd committed:
Merge pull request #13656 from chanks/rollback_transactions_in_killed_threads

Data corruption risk: Roll back open transactions when the running thread is killed.

spastorino committed:
Merge pull request #16910 from kuldeepaggarwal/use-rails-options

use `allow_blank` option instead of custom condition

yuki24 committed:
Rescue Rack::Utils::ParameterTypeError instead of TypeError

As of rack/rack@167b6480235ff00ed5f355698bf00ec2f250f72e, Rack raises
Rack::Utils::ParameterTypeError which inherits TypeError.

In terms of the behavior, Rescuing TypeError still works but this
method shouldn't rescue if TypeError is raised for other reasons.

jeremy committed:
Merge pull request #16758 from Agis-/time-change-nsec

Add support for `:nsec` option to `Time#change`

jeremy committed:
Merge pull request #16912 from yuki24/rescue-rack-parameter-type-error

Rescue `Rack::Utils::ParameterTypeError` instead of `TypeError`

jeremy committed:
Tighten up AJ::Arguments and its tests

* Disallow deserialization of non-primitive objects
* Broaden coverage; remove superfluous tests

jeremy committed:
Pare down and simplify Active Job's rake tasks

kiuru committed:
Singular credit_card controller name changed to plural

tenderlove committed:
Stop nil checking on safe_append=

ERB compiler guarantees safe_append= will be called with a string, so
nil checks don't make sense.  Anything else calling this method should
check for nil themselves before calling

seuros committed:
simplify test_adapter

seuros committed:
Merge pull request #16918 from kiuru/master

Singular credit_card controller name changed to plural [ci skip]

jeremy committed:
Merge pull request #16920 from seuros/aj

simplify test_adapter

y-yagi committed:
Add `test:jobs` task

tenderlove committed:
handle <%== nil %> cases

This is much less common than string literal appends, so add a special
case method for it.  Maybe fixes bug reported by @jeremy on 97ef636191933f1d4abc92fc10871e6d1195285c

chancancode committed:
Merge pull request #16922 from y-yagi/add_test_jobs_task

Add test jobs task

jeremy committed:
Merge pull request #16574 from lsylvester/duration-respond_to

add implementation of respond_to? for ActiveSupport::Duration

alan committed:
Don't autosave unchanged has_one through records

matthewd committed:
As of Unicode 6.3, Mongolian Vowel Separator is not whitespace

Ruby 2.2 knows this, and no longer matches it with [[:space:]], so it's
not a good candidate for testing String#squish.

greysteil commented:
@alan - you should close this, it's on your own repo :)
